system/trace/trace.c

```diff
@@ -0,0 +1,812 @@
+
+  /* Parse the setting parameters */
+
+  while (index < argc - 1)
+    {
+      if (argv[index + 1][0] != '-' && argv[index + 1][0] != '+')
+        {
+          break;
+        }
+
+      index++;
+      modified = true;
+      enable = (argv[index][0] == '+');
+
+      if (argv[index][1] == '*')
+        {
+          /* Mask or unmask all IRQs */
+
+          if (enable)
+            {
+              for (n = 0; n < NR_IRQS; n++)
+                {
+                  NOTE_FILTER_IRQMASK_SET(n, &filter_irq);
+                }
+            }
+          else
+            {
+              NOTE_FILTER_IRQMASK_ZERO(&filter_irq);
+            }
+
+          continue;
+        }
+
+      /* Get IRQ number */
+
+      irqno = strtoul(&argv[index][1], &endptr, 0);
+      if (endptr == &argv[index][1] || *endptr != '\0' ||
+          irqno >= NR_IRQS)
+        {
+          fprintf(stderr,
+                  "trace irq: invalid argument '%s'\n", argv[index]);
+          return ERROR;
+        }
+
+      /* Update the masked IRQ number list */
+
+      if (enable)
+        {
+          NOTE_FILTER_IRQMASK_SET(irqno, &filter_irq);
+        }
+      else
+        {
+          NOTE_FILTER_IRQMASK_CLR(irqno, &filter_irq);
+        }
+    }
+
+  if (modified)
+    {
+      /* Update current irq filter setting */
+
+      ioctl(g_notectlfd, NOTECTL_SETIRQFILTER, (unsigned long)&filter_irq);
+
+      /* Enable irq trace flag */
+
+      ioctl(g_notectlfd, NOTECTL_GETMODE, (unsigned long)&mode);
+      mode.flag |= NOTE_FILTER_MODE_FLAG_IRQ;
+      ioctl(g_notectlfd, NOTECTL_SETMODE, (unsigned long)&mode);
+    }
+  else
+    {
+      /* If no parameter, display current setting. */
+
+      count = 0;
+      for (n = 0; n < NR_IRQS; n++)
+        {
+          if (NOTE_FILTER_IRQMASK_ISSET(n, &filter_irq))
+            {
+              count++;
+            }
+        }
+
+      printf("Filtered IRQs: %d\n", count);
+
+      for (n = 0; n < NR_IRQS; n++)
+        {
+          if (NOTE_FILTER_IRQMASK_ISSET(n, &filter_irq))
+            {
+              printf("  %d\n", n);
+            }
+        }
+    }
+
+  return index;
+}
+#endif
+
+/****************************************************************************
+ * Name: show_usage
+ ****************************************************************************/
+
+static void show_usage(FAR const char *progname, int exitcode)
+{
+  fprintf(stderr,
+          "\nUsage: trace <subcommand>...\n"
+          "Subcommand:\n"
+          "  start [<duration>]              :"
+                                " Start task tracing\n"
+          "  stop                            :"
+                                " Stop task tracing\n"
+#ifdef CONFIG_SYSTEM_SYSTEM
+          "  cmd <command> [<args>...]       :"
+                                " Get the trace while running <command>\n"
+#endif
+#ifdef CONFIG_DRIVER_NOTERAM
+          "  dump [<filename>]               :"
```

Comment:
Since trace without argument is same as trace dump wihtout filename, how about we move {+|-}o to here to centralize all dump option in one place? e.g.:
dump ->show the overwrite state
dump filename->save to file
dump +o->overwrite
dump -o->don't ovwerwrite
The benefit is that, we can add function to trace_dump.c like:
trace_dump_cmd
trace_dump_show_usage
and trace.c just forward to these function.

Reply:
I tried to change trace_cmd_dump() to move overwrite option from trace_cmd_mode().
But, I think it should not be moved into trace_dump.c. Because,
- The command line parsing is split into two files.
- trace_cmd_dump() needs to call notectl_enable() because https://github.com/apache/incubator-nuttx-apps/pull/421#discussion_r502805629 . Then, both files need to call another mutually and it becomes not simple.

Reply:
Though once I have tried to fix "trace dump" command to follow your proposals, I have reverted the changes.
Especially, the following syntax seems strange to me.
trace -> dump trace data to console
trace dump ->show the overwrite state
trace dump filename->save to file

I have reverted it and changed the default behavior of trace command (without options).
trace -> show current state (same as "trace mode")
trace dump -> dump trace data to console
trace dump filename -> save to file (dump trace data to file)
How about it?

Reply:
No problem.